### PR TITLE
Bug 1281901 - Swift 2.3 Migration

### PR DESCRIPTION
**Pull request by sleroux**
Opening PR for visibility into WIP.

### CODE REVIEW

**Comment by sleroux** [comment]
TL:DR; Need some more time to investigate converting of old version of our dependencies to Swift 2.3.

I dove back into this today to see what it would take to land this and get us working on Xcode 8 when it comes out soon and discovered some big issues. Good news is that most of our dependencies have either updated their mainline release to Swift 2.3 or have branches/releases specifically supporting it. Bad news is that because we haven't updated our dependencies in a year or so, some of these libraries have updated their APIs and the Swift 2.3-ported version of the library is the latest-and-greatest.

For example, two libraries I've discovered so far that have updated their APIs since we've updated include Deferred (https://github.com/bignerdranch/Deferred/releases) and Alamofire (https://github.com/Alamofire/Alamofire/releases). Taking the older version of Deferred and converting it to Swift 2.3 doesn't require any source changes but I haven't tried it on Alamofire yet which might have issues considering its a pretty large library. I'll continue to investigate and see if we can get by with taking our older snapshots of our dependencies and converting them to Swift 2.3 just so we can use Xcode 8.

Long term, this is really bad. We need to make sure we keep up to date with these library changes considering changing landscape of Swift. Alternatively we can find Objective-C dependencies instead to avoid binary compatibility issues but that seems like a step backwards. We should probably also clean up where we are getting our dependencies from. Some of our libraries are coming from open branches on people's forks of the original project which isn't a big deal but we should find a way to standardize and organize this (maybe put them under the Mozilla org?).

**Comment by sleroux** [comment]
I've updated the Cartfile to point to Swift 2.3 compatible dependencies. As for Alamofire and Deferred,

* Alamofire 3.5.0 as some API changes from the current version we're using (2.0.2). From what I can tell, the only change is in the completion handler closures and how it passes the response data back. Instead of 3 values, its now just 1 response object containing these three values. We'll need to go through each of our targets and make the updates.

* For Deferred, I forked the existing fork @fluffyemily had and made a small commit to make the framework that it spits out compatible with Swift 2.3. Since the code didn't need any changes, Deferred should just work without any updates to our codebase. We should eventually look to update though.

After the Alamofire updates we should be good to go for now in Swift 2.3!

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -3233,9 +3235,9 @@ class BlurWrapper: UIView {
             }
             effectView.hidden = disableBlur
             switch blurStyle {
-            case .ExtraLight, .Light:
+            case .ExtraLight, .Light, .Regular, .Prominent:
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: I'm not sure what these `Regular` and `Prominent` styles are, and I can't find much documentation about them. I arbitrarily bucketed them into the light style, but let me know if you can figure out what these are...

Comment (sleroux) [inline Client/Frontend/Browser/BrowserViewController.swift]: From a transcipt of a WWDC talk [1]

> So if you use UIBlurEffectStyleRegular and the system's in light, it will use UIBlurEffectStyle.light.
> If you use regular and dark, you'll use dark.
> If you use prominent, it will use .extraLight and .extraDark. 

I think for our uses in this method we want to ignore the system blur and just stick with what we were doing.

[1] http://asciiwwdc.com/2016/sessions/206

File: Client/Frontend/Share/ShareExtensionHelper.swift [hunk 2]
```diff
@@ -25,7 +25,7 @@ class ShareExtensionHelper: NSObject {
         var activityItems = [AnyObject]()
 
         let printInfo = UIPrintInfo(dictionary: nil)
-        printInfo.jobName = selectedTab?.url?.absoluteString ?? selectedURL.absoluteString
+        printInfo.jobName = selectedTab?.url?.absoluteString ?? selectedURL.absoluteString!
```
Comment (sleroux) [inline Client/Frontend/Share/ShareExtensionHelper.swift]: Nit: Might be safer to guard here. Not sure if it matters though.

File: Sync/RequestExtensions.swift [hunk 3]
```diff
@@ -7,77 +7,65 @@ import Alamofire
...
-    public func responseParsedJSON(queue queue: dispatch_queue_t, completionHandler: ResponseHandler) -> Self {
-        let serializer = ParsedJSONResponseSerializer()
-        return self.response(queue: queue, responseSerializer: serializer, completionHandler: completionHandler)
+    public func responseParsedJSON(queue queue: dispatch_queue_t, completionHandler: Response<JSON, JSONSerializeError> -> ()) -> Self {
+        return self.response(queue: queue, responseSerializer: parsedJSONResponseSerializer(), completionHandler: completionHandler)
     }
 }
 
-private enum JSONSerializeError: ErrorType {
+public enum JSONSerializeError: ErrorType {
```
Comment (sleroux) [inline Sync/RequestExtensions.swift]: How come we need to make this private now?

Comment (sleroux) [inline Sync/RequestExtensions.swift]: Nevermind! I see it gets used above in the methods as a parameter value.

**Comment by farhanpatel** [comment]
http://i.giphy.com/L9kwXOK00OJGM.gif
